src/sparse_pattern.jl

```diff
@@ -0,0 +1,109 @@
+"""
+    p_norm_sparsity_matrix(M::AbstractArray, ratio::Real, p::Real, min_per_row=0::Integer, min_per_col=0::Integer)
+
+Compute a pattern matrix `M_pat` which is a p-norm sparsity pattern for `M`.
+
+`M_pat` is a sparse `SparseMatrixCSC{Int64, Int64}` of the same shape as `M` and
+contains only 0 or 1.
+
+`ratio` should be in ``[0,1]``.  `p` should be in ``(0, \\infty ]``
+
+* `min_per_row` is the minimum number of non-zeros needed per row.
+* `min_per_col` is the minimum number of non-zeros needed per column.
+
+Defaults are `0`.
+
+See also: [`sps_compute`](@ref), [`bin_sparse_matrix!`](@ref).
+
+# Examples
+```jldoctest
+julia> p_norm_sparsity_matrix([4 1; 0.1 17], 0.6, 2)
+2×2 SparseMatrixCSC{Int64, Int64} with 2 stored entries:
+ 1  ⋅
+ ⋅  1
+```
+"""
+function p_norm_sparsity_matrix(M::AbstractArray, ratio::Real, p::Real, min_per_row=0::Integer, min_per_col=0::Integer)
+    @assert 0 <= ratio <= 1 "ration must be in [0,1]"
```

Comment:
I think it's better to throw errors than to raise asserts: https://docs.julialang.org/en/v1/base/base/#Base.@assert

When checking whether user-defined arguments are valid, you can use `ArgumentError` or `DomainError`: 
```julia
0 <= ratio <= 1 || throw(DomainError(ratio, "ratio must be in [0,1]"))
```
See also: https://docs.julialang.org/en/v1/base/base/#Errors